src/rust/engine/ui/src/console_ui.rs

```diff
@@ -78,18 +77,33 @@ impl ConsoleUI {
   ///
   /// Setup progress bars, and return them along with a running task that will drive them.
   ///
+  /// NB: This method must be very careful to avoid logging. Between the point where we have taken
+  /// exclusive access to the Console and when the UI has actually been initialized, attempts to
+  /// log from this method would deadlock (by causing the method to wait for _itself_ to finish).
+  ///
   fn setup_bars(
     &self,
     executor: Executor,
   ) -> Result<(Vec<ProgressBar>, MultiProgressTask), String> {
-    // Stderr is propagated across a channel to remove lock interleavings between stdio and the UI.
```

Comment:
After giving this some thought, I believe that the lock interleavings referred to here are no longer possible as the code exists today. The `indicatif` `ProgressBar` has an internal `RwLock`, but none of our code executes underneath it (because it accesses `stdio` directly while it has exclusive access). So it will never end up "above" one of our own locks in the callstack.